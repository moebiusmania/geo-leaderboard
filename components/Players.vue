<script setup lang="ts">
const { entries } = defineProps<{ entries: Entry[] }>();

interface Player extends Entry {
  avatar: string;
  level: number;
}

const avatars = [
  "https://cdn.midjourney.com/793e43a4-d703-450f-8172-584d6ef688f5/0_1.png",
  `https://cdn.midjourney.com/26514a12-132b-464d-87b4-865cc9934d8b/0_3.png`,
];

// convert milliseconds to minutes and seconds
const formatTime = (time: number): string => {
  const minutes = Math.floor(time / 60000);
  const seconds = ((time % 60000) / 1000).toFixed(0);
  return `${minutes}:${seconds.padStart(2, "0")}`;
};

const getLevel = (score: number): number => {
  return Math.floor(score / 20);
};

const players: Player[] = entries.map((entry) => ({
  ...entry,
  avatar: avatars[Math.floor(Math.random() * avatars.length)],
  level: getLevel(entry.Score),
}));
</script>

<template>
  <section>
    <ul>
      <li v-for="player in players" :key="player.EntryId">
        <img :src="player.avatar" alt="Player avatar" />
        <span class="name">{{ player.Name }}</span>
        <span class="score">{{ player.Score }}xp</span>
        <Ribbon :value="player.level" />
        <Progress :current="player.level" />
      </li>
    </ul>
  </section>
</template>

<style scoped>
section {
  margin-top: 5rem;
  display: flex;
  justify-content: center;

  & ul {
    display: flex;
    flex-direction: column;
    gap: 1rem;
    margin: 0;
    padding: 0;
    max-width: 900px;
    width: 100%;

    & li {
      list-style: none;
      padding: 0;
      display: flex;
      justify-content: space-between;
      align-items: center;
      gap: 1rem;

      & img {
        width: 45px;
        height: 45px;
        border-radius: 6px;
        object-fit: cover;
        border: 5px solid var(--primary);
      }

      & span {
        font-size: 1.2rem;
        font-weight: 600;
      }

      & .name {
        font-family: var(--title-font);
        font-size: 1.5rem;
        min-width: 100px;
      }

      & .score {
        min-width: 60px;
      }
    }
  }
}
</style>
