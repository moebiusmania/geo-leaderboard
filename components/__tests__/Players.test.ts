import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import Players from "../Players.vue";

describe("Players", () => {
  const mockSeason = {
    number: 1,
    title: "Test Season",
    levels: 5,
    cap: 1000,
    homeworks: [],
  };

  const mockPlayers = [
    {
      _type: "Entry",
      EntryId: "1",
      Name: "Test Player 1",
      Score: 750,
      Time: 1000,
      gender: "male",
    },
    {
      _type: "Entry",
      EntryId: "2",
      Name: "Test Player 2",
      Score: 250,
      Time: 1000,
      gender: "female",
    },
  ];

  it("renders properly with players data", () => {
    const wrapper = mount(Players, {
      props: {
        data: mockPlayers,
        season: mockSeason,
      },
    });

    // Check if all players are rendered
    const playerItems = wrapper.findAll("li");
    expect(playerItems).toHaveLength(2);

    // Check if player names are displayed
    expect(wrapper.text()).toContain("Test Player 1");
    expect(wrapper.text()).toContain("Test Player 2");

    // Check if scores are displayed
    expect(wrapper.text()).toContain("750");
    expect(wrapper.text()).toContain("250");

    // Create snapshot
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("calculates progress and levels correctly", () => {
    const wrapper = mount(Players, {
      props: {
        data: mockPlayers,
        season: mockSeason,
      },
    });

    // Test progress calculation (750/1000 * 100 = 75%)
    const firstPlayer = wrapper.findAll("li")[0];
    const progress = firstPlayer.findComponent({ name: "Progress" });
    expect(progress.props("progress")).toBe(75);

    // Test level calculation (75% in 5 levels should be level 4)
    const ribbon = firstPlayer.findComponent({ name: "Ribbon" });
    expect(ribbon.props("value")).toBe(4);
  });
});
