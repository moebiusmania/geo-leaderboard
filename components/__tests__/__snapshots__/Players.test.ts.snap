// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Players > renders properly with players data 1`] = `
"<section data-v-d41cfe1f="">
  <ul data-v-d41cfe1f="">
    <li data-v-d41cfe1f=""><img data-v-d41cfe1f="" src="https://cdn.midjourney.com/26514a12-132b-464d-87b4-865cc9934d8b/0_3.png" alt="Player avatar"><span data-v-d41cfe1f="" class="name">Test Player 1</span><span data-v-d41cfe1f="" class="score">750<span data-v-d41cfe1f="">xp</span></span><span data-v-dc7e820b="" data-v-d41cfe1f="" class="ribbon animate__animated animate__jello"> lvl<span data-v-dc7e820b="">4</span></span>
      <div data-v-0b11ee78="" data-v-d41cfe1f="" class="progress-container">
        <div data-v-0b11ee78="" class="progress-bar">
          <div data-v-0b11ee78="" class="progress-fill" style="width: 75%;"></div>
          <div data-v-0b11ee78="" class="circles">
            <div data-v-0b11ee78="" class="circle active"></div>
            <div data-v-0b11ee78="" class="circle active"></div>
            <div data-v-0b11ee78="" class="circle active"></div>
            <div data-v-0b11ee78="" class="circle active"></div>
            <div data-v-0b11ee78="" class="circle"></div>
          </div>
        </div>
      </div>
    </li>
    <li data-v-d41cfe1f=""><img data-v-d41cfe1f="" src="https://cdn.midjourney.com/793e43a4-d703-450f-8172-584d6ef688f5/0_1.png" alt="Player avatar"><span data-v-d41cfe1f="" class="name">Test Player 2</span><span data-v-d41cfe1f="" class="score">250<span data-v-d41cfe1f="">xp</span></span><span data-v-dc7e820b="" data-v-d41cfe1f="" class="ribbon animate__animated animate__jello"> lvl<span data-v-dc7e820b="">2</span></span>
      <div data-v-0b11ee78="" data-v-d41cfe1f="" class="progress-container">
        <div data-v-0b11ee78="" class="progress-bar">
          <div data-v-0b11ee78="" class="progress-fill" style="width: 25%;"></div>
          <div data-v-0b11ee78="" class="circles">
            <div data-v-0b11ee78="" class="circle active"></div>
            <div data-v-0b11ee78="" class="circle active"></div>
            <div data-v-0b11ee78="" class="circle"></div>
            <div data-v-0b11ee78="" class="circle"></div>
            <div data-v-0b11ee78="" class="circle"></div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</section>"
`;
